File: internal/services/urls.go
package services

import (
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"unicode"
)

var navPathCharset = regexp.MustCompile(`^[A-Za-z0-9/_#?&=%.~+-]+$`)

// ValidateHTTPSURL checks that raw is empty (when allowed) or an https URL.
// Rejects javascript:, data:, vbscript:, and scheme-relative //… forms.
func ValidateHTTPSURL(raw string, allowEmpty bool) error {
	s := strings.TrimSpace(raw)
	if s == "" {
		if allowEmpty {
			return nil
		}
		return fmt.Errorf("%w: url is required", ErrValidation)
	}
	if strings.HasPrefix(s, "//") {
		return fmt.Errorf("%w: scheme-relative urls are not allowed", ErrValidation)
	}
	lower := strings.ToLower(s)
	for _, bad := range []string{"javascript:", "data:", "vbscript:"} {
		if strings.HasPrefix(lower, bad) {
			return fmt.Errorf("%w: disallowed url scheme", ErrValidation)
		}
	}
	u, err := url.Parse(s)
	if err != nil {
		return fmt.Errorf("%w: invalid url", ErrValidation)
	}
	if !strings.EqualFold(u.Scheme, "https") {
		return fmt.Errorf("%w: url must use https", ErrValidation)
	}
	return nil
}

// ValidateNavPath checks that raw is a non-empty relative path starting with a
// single /, with no scheme and an allowlisted character set.
func ValidateNavPath(raw string) error {
	s := strings.TrimSpace(raw)
	if s == "" {
		return fmt.Errorf("%w: path is required", ErrValidation)
	}
	if strings.HasPrefix(s, "//") {
		return fmt.Errorf("%w: path must not be scheme-relative", ErrValidation)
	}
	if !strings.HasPrefix(s, "/") {
		return fmt.Errorf("%w: path must start with /", ErrValidation)
	}
	if strings.Contains(s, "://") {
		return fmt.Errorf("%w: path must be relative", ErrValidation)
	}
	if !navPathCharset.MatchString(s) {
		return fmt.Errorf("%w: path contains invalid characters", ErrValidation)
	}
	return nil
}

// ValidateEmail allows empty values. Non-empty values must contain @ and must
// not include whitespace, colon, or control characters.
func ValidateEmail(raw string) error {
	s := strings.TrimSpace(raw)
	if s == "" {
		return nil
	}
	if strings.HasPrefix(s, "//") {
		return fmt.Errorf("%w: invalid email", ErrValidation)
	}
	if !strings.Contains(s, "@") {
		return fmt.Errorf("%w: invalid email", ErrValidation)
	}
	for _, r := range s {
		if r == ':' || unicode.IsSpace(r) || unicode.IsControl(r) {
			return fmt.Errorf("%w: invalid email", ErrValidation)
		}
	}
	return nil
}

File: web/src/lib/urls.ts
const NAV_PATH_RE = /^[A-Za-z0-9/_#?&=%.~+-]+$/;

/** True when raw is empty (if allowed) or an https URL. Rejects javascript:/data:/vbscript:/ //. */
export function isValidHTTPSURL(raw: string, allowEmpty = false): boolean {
  const s = raw.trim();
  if (!s) return allowEmpty;
  if (s.startsWith("//")) return false;
  const lower = s.toLowerCase();
  if (
    lower.startsWith("javascript:") ||
    lower.startsWith("data:") ||
    lower.startsWith("vbscript:")
  ) {
    return false;
  }
  try {
    const u = new URL(s);
    return u.protocol.toLowerCase() === "https:";
  } catch {
    return false;
  }
}

/** True when raw is a relative path starting with a single /. */
export function isValidNavPath(raw: string): boolean {
  const s = raw.trim();
  if (!s || s.startsWith("//") || !s.startsWith("/")) return false;
  if (s.includes("://")) return false;
  return NAV_PATH_RE.test(s);
}

/** Empty is valid. Non-empty must contain @ and no whitespace, colon, or control chars. */
export function isValidEmail(raw: string): boolean {
  const s = raw.trim();
  if (!s) return true;
  if (s.startsWith("//")) return false;
  if (!s.includes("@")) return false;
  for (const ch of s) {
    const code = ch.charCodeAt(0);
    if (ch === ":" || code <= 32 || code === 127) return false;
  }
  return true;
}

/** Returns a trimmed https URL, or undefined when unsafe/empty. */
export function safeHref(raw: string | null | undefined): string | undefined {
  if (raw == null) return undefined;
  const s = raw.trim();
  if (!s || !isValidHTTPSURL(s, false)) return undefined;
  return s;
}

/** Returns a trimmed nav path, or undefined when unsafe. */
export function safePath(raw: string | null | undefined): string | undefined {
  if (raw == null) return undefined;
  const s = raw.trim();
  if (!isValidNavPath(s)) return undefined;
  return s;
}
